## Importing the Dependencies

In [30]:
import pandas as pd
from pathlib import Path
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import *
import joblib
import seaborn as sns

## Data Collection And Preprocessing

In [2]:
# Load the dataset in a variable named 'hd' using Pandas DataFrame
data_path = Path("../data/heart_disease.csv")
hd = pd.read_csv(data_path)

In [3]:
# Print first 5 rows of the dataset
hd.head()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
0,52,1,0,125,212,0,1,168,0,1.0,2,2,3,0
1,53,1,0,140,203,1,0,155,1,3.1,0,0,3,0
2,70,1,0,145,174,0,1,125,1,2.6,0,0,3,0
3,61,1,0,148,203,0,1,161,0,0.0,2,1,3,0
4,62,0,0,138,294,1,1,106,0,1.9,1,3,2,0


In [4]:
# Print Last 5 rows of the dataset
hd.tail()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
4841,63,1,2,174,281,0,0,154,0,0.848749,1,0,3,0
4842,59,1,1,140,184,0,0,155,0,2.804466,1,0,2,0
4843,53,1,0,123,282,0,1,95,1,2.000000,1,2,3,0
4844,58,1,1,112,230,0,0,164,0,2.326788,1,1,3,0
4845,66,1,0,120,239,0,0,78,0,1.363850,0,0,2,0


In [5]:
# Print the rows and columns of the dataset
hd.shape


(4846, 14)

There are 4846 records and 14 features in the dataset

In [6]:
# Getting some information about the features in the dataset
hd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4846 entries, 0 to 4845
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       4846 non-null   int64  
 1   sex       4846 non-null   int64  
 2   cp        4846 non-null   int64  
 3   trestbps  4846 non-null   int64  
 4   chol      4846 non-null   int64  
 5   fbs       4846 non-null   int64  
 6   restecg   4846 non-null   int64  
 7   thalach   4846 non-null   int64  
 8   exang     4846 non-null   int64  
 9   oldpeak   4846 non-null   float64
 10  slope     4846 non-null   int64  
 11  ca        4846 non-null   int64  
 12  thal      4846 non-null   int64  
 13  target    4846 non-null   int64  
dtypes: float64(1), int64(13)
memory usage: 530.2 KB


In [7]:
# Statistical measures of the data
hd.describe()

,age,sex,cp,trestbps,chol,fbs,restecg,thalach,exang,oldpeak,slope,ca,thal,target
count,4846.000000,4846.00000,4846.000000,4846.000000,4846.000000,4846.000000,4846.000000,4846.000000,4846.000000,4846.000000,4846.000000,4846.000000,4846.000000,4846.000000
mean,56.097400,0.76558,0.498349,133.448411,249.419934,0.116178,0.369170,140.742262,0.409410,1.512251,1.111019,0.913950,2.403013,0.108543
std,7.746713,0.42368,0.862651,17.655262,50.224019,0.320472,0.507234,22.938195,0.491776,1.225896,0.583621,0.957628,0.686283,0.311097
min,29.000000,0.00000,0.000000,94.000000,126.000000,0.000000,0.000000,71.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,52.000000,1.00000,0.000000,120.000000,212.000000,0.000000,0.000000,125.000000,0.000000,0.525232,1.000000,0.000000,2.000000,0.000000
50%,57.000000,1.00000,0.000000,130.000000,246.000000,0.000000,0.000000,143.000000,0.000000,1.400000,1.000000,1.000000,3.000000,0.000000
75%,61.000000,1.00000,1.000000,142.000000,283.000000,0.000000,1.000000,159.000000,1.000000,2.200000,1.000000,2.000000,3.000000,0.000000
max,77.000000,1.00000,3.000000,200.000000,564.000000,1.000000,2.000000,202.000000,1.000000,6.200000,2.000000,4.000000,3.000000,1.000000


In [8]:
# Printing the column names
hd.columns

Index(['age', 'sex', 'cp', 'trestbps', 'chol', 'fbs', 'restecg', 'thalach',
       'exang', 'oldpeak', 'slope', 'ca', 'thal', 'target'],
      dtype='object')

Target is the target Column


In [9]:
# Distribution of target Variable
hd['target'].value_counts()

target
0    4320
1     526
Name: count, dtype: int64

0-> No Heart Disease
1-> Heart Disease Exists

In [10]:
# Checking for the missing values
hd.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

No missing values are present in the dataset

## Train-Test-Split

In [11]:
# Creating independent and dependent variable
X = hd.drop("target", axis=1)
y = hd["target"]


In [12]:
# Splitting the data for training 80% and for testing 20%
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [13]:
# Scaling the training values
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

## Logistic Regression

In [14]:
# Training with the logistic Model
lm=LogisticRegression(class_weight="balanced")
lm.fit(X_train, y_train)

LogisticRegression(class_weight='balanced')

In [15]:
# Predicting the testing data
y_pred = lm.predict(X_test)

In [16]:
# Getting the Accuracy Score
accuracy_score(y_test, y_pred)

0.8484536082474227

In [17]:
# Printing the Classification Report
print(classification_report(y_test, y_pred))

              precision    recall  f1-score   support

           0       0.99      0.84      0.91       865
           1       0.41      0.91      0.57       105

    accuracy                           0.85       970
   macro avg       0.70      0.88      0.74       970
weighted avg       0.93      0.85      0.87       970



In [18]:
# Printing the Confusion Matrix
confusion_matrix(y_test, y_pred)

array([[727, 138],
       [  9,  96]])

## Random Forest

In [19]:
# Training with the Random forest Model
rf = RandomForestClassifier(random_state=42)
rf.fit(X_train, y_train)

RandomForestClassifier(random_state=42)

In [20]:
# Predicting the testing data
rf_pred = rf.predict(X_test)

In [21]:
# Getting the Accuracy Score
accuracy_score(y_test, rf_pred)

0.9989690721649485

In [22]:
# Printing the Classification Report
print(classification_report(y_test, rf_pred))

              precision    recall  f1-score   support

           0       1.00      1.00      1.00       865
           1       0.99      1.00      1.00       105

    accuracy                           1.00       970
   macro avg       1.00      1.00      1.00       970
weighted avg       1.00      1.00      1.00       970



In [23]:
# Printing the Confusion Matrix
confusion_matrix(y_test, rf_pred)

array([[864,   1],
       [  0, 105]])

In [24]:
# Getting the features by their contribution in predicting the Target
importance = pd.Series(rf.feature_importances_, index=X.columns)
importance.sort_values(ascending=False)

thalach     0.130066
chol        0.122690
age         0.118552
cp          0.112825
trestbps    0.106863
oldpeak     0.106563
slope       0.074800
ca          0.070253
sex         0.047458
thal        0.043781
restecg     0.030899
exang       0.019957
fbs         0.015292
dtype: float64

## Model Comparison and Selection

Random Forest achieved the best performance with the highest
accuracy and recall, making it the selected model for heart
disease prediction.

In [29]:
# Saving the Pickle file
joblib.dump(rf, "../models/heart_disease_model.pkl")

['../models/heart_disease_model.pkl']

## Predictive Model

In [26]:
def predict_heart_disease(input_data):

    import numpy as np

    input_data = np.array(input_data)
    input_data = input_data.reshape(1, -1)

    prediction = rf.predict(input_data)

    if prediction[0] == 1:
        return "Heart Disease Detected"
    else:
        return "No Heart Disease"

In [27]:
sample_patient = [63, 1, 3, 145, 233, 1, 0, 150, 0, 2.3, 0, 0, 1]

In [28]:
predict_heart_disease(sample_patient)

'No Heart Disease'